lnd.go

```diff
@@ -1633,102 +1624,87 @@ func initializeDatabases(ctx context.Context,
-
-		case err != nil:
-			err := fmt.Errorf("unable to open local channeldb: %v", err)
-			ltndLog.Error(err)
-			return nil, nil, err
+		cleanUp = func() {
+			for name, closeFunc := range closeFuncs {
+				if err := closeFunc(); err != nil {
+					ltndLog.Errorf("Error closing %s "+
+						"database: %v", name, err)
+				}
+			}
 		}
+	)
+	if databaseBackends.Remote {
+		ltndLog.Infof("Using remote %v database! Creating "+
+			"graph and channel state DB instances", cfg.DB.Backend)
+	} else {
+		ltndLog.Infof("Creating local graph and channel state DB " +
+			"instances")
+	}
+
+	// Otherwise, we'll open two instances, one for the state we only need
+	// locally, and the other for things we want to ensure are replicated.
+	dbs.graphDB, err = channeldb.CreateWithBackend(
+		databaseBackends.GraphDB,
+		channeldb.OptionSetRejectCacheSize(cfg.Caches.RejectCacheSize),
+		channeldb.OptionSetChannelCacheSize(cfg.Caches.ChannelCacheSize),
+		channeldb.OptionSetBatchCommitInterval(cfg.DB.BatchCommitInterval),
+		channeldb.OptionDryRunMigration(cfg.DryRunMigration),
+	)
+	switch {
+	// Give the DB a chance to dry run the migration. Since we know that
+	// both the channel state and graph DBs are still always behind the same
+	// backend, we know this would be applied to both of those DBs.
+	case err == channeldb.ErrDryRunMigrationOK:
+		ltndLog.Infof("Graph DB dry run migration successful")
+		return nil, nil, err
 
-		closeFuncs = append(closeFuncs, func() {
-			dbs.graphDB.Close()
-		})
-
-		ltndLog.Infof("Opening replicated database instance...")
-
-		dbs.chanStateDB, err = channeldb.CreateWithBackend(
-			databaseBackends.ChanStateDB,
-			channeldb.OptionDryRunMigration(cfg.DryRunMigration),
-			channeldb.OptionSetBatchCommitInterval(cfg.DB.BatchCommitInterval),
-		)
-		switch {
-		case err == channeldb.ErrDryRunMigrationOK:
-			return nil, nil, err
+	case err != nil:
+		cleanUp()
 
-		case err != nil:
-			dbs.graphDB.Close()
+		err := fmt.Errorf("unable to open graph DB: %v", err)
+		ltndLog.Error(err)
+		return nil, nil, err
+	}
 
-			err := fmt.Errorf("unable to open remote channeldb: %v", err)
-			ltndLog.Error(err)
-			return nil, nil, err
-		}
+	// For now, we don't _actually_ split the graph and channel state DBs on
+	// the code level. Since they both are based upon the *channeldb.DB
+	// struct it will require more refactoring to fully separate them. With
+	// the full remote mode we at least know for now that they both point to
+	// the same DB backend (and also namespace within that) so we only need
+	// to apply any migration once.
+	//
+	// TODO(guggero): Remove once full separation between graph and channel
+	// state DB is done.
+	dbs.chanStateDB = dbs.graphDB
 
-		closeFuncs = append(closeFuncs, func() {
-			dbs.chanStateDB.Close()
-		})
-	}
+	// Success, let's overwrite the close function with the decorated one.
+	closeFuncs[lblGraphDB] = dbs.graphDB.Close
+	closeFuncs[lblStateDB] = dbs.chanStateDB.Close
```

Comment:
nit: calling the `Close` multiple times may result in error logs.

Reply:
Oops, this was a left over rebase artifact. Each backend's `Close()` function should now only be called once.